File: src/40.cpp
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int>> res;
        
        if (candidates.size() == 0) {
            return res;
        }
        vector<int> temp;
        sort(candidates.begin(), candidates.end());
        dfs(candidates, target, res, 0, temp);
        return res;
        
    }
    
    void dfs(vector<int> &candidates, int target, vector<vector<int>> &res, int index, vector<int> temp) {
        if (target < 0) {
            return;
        }
        if (target == 0) {
            res.push_back(temp);
        } else {
            for (int i = index; i < candidates.size(); i ++) {
                if (i > index && candidates[i] == candidates[i - 1]) {
                    continue;
                }
                temp.push_back(candidates[i]);
                dfs(candidates, target - candidates[i], res, i + 1, temp);
                temp.pop_back();
            }
        }
    }
};

File: src/40.py
class Solution:
    """
    @param candidates: A list of integers
    @param target: An integer
    @return: A list of lists of integers
    """
    def combinationSum2(self, candidates, target):
        # write your code here
        results = []
        if candidates == None:
            return results
        candidates.sort()
        self.dfs(candidates, 0, target, [], results)
        return results
    #递归的定义
    #找到所有以combanation 开头 和为target的组合 并放入result， 其中剩余需要加入combanation数的和为remainTarget
    #并且下一个加入的位置从startindex开始
    def dfs(self, candidates, startIndex, remainTarget, combination, results):
        if remainTarget < 0:
            return

        if remainTarget == 0:
            results.append([] + combination)
        else:
            for i in range(startIndex, len(candidates)):
                if i > startIndex and candidates[i] == candidates[i - 1]:
                    continue
                combination.append(candidates[i])
                self.dfs(candidates,i + 1,remainTarget - candidates[i], combination, results)
                combination.pop()

            

